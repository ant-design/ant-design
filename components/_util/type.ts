import type React from 'react';

/** https://github.com/Microsoft/TypeScript/issues/29729 */
export type LiteralUnion<T extends string> = T | (string & {});

export type AnyObject = Record<PropertyKey, any>;

export type CustomComponent<P = AnyObject> = React.ComponentType<P> | string;

/**
 * Get component props
 * @example
 * ```ts
 * import { Checkbox } from 'antd'
 * import type { GetProps } from 'antd';
 *
 * type CheckboxGroupProps = GetProps<typeof Checkbox.Group>
 * ```
 * @since 5.13.0
 */
export type GetProps<T extends React.ComponentType<any> | object> = T extends React.ComponentType<
  infer P
>
  ? P
  : T extends object
    ? T
    : never;

/**
 * Get component props by component name
 * @example
 * ```ts
 * import { Select } from 'antd';
 * import type { GetProp, SelectProps } from 'antd';
 *
 * type SelectOption1 = GetProp<SelectProps, 'options'>[number];
 * // or
 * type SelectOption2 = GetProp<typeof Select, 'options'>[number];
 *
 * const onChange: GetProp<typeof Select, 'onChange'> = (value, option) => {
 *  // Do something
 * };
 * ```
 * @since 5.13.0
 */
export type GetProp<
  T extends React.ComponentType<any> | object,
  PropName extends keyof GetProps<T>,
> = NonNullable<GetProps<T>[PropName]>;

type ReactRefComponent<Props extends { ref?: React.Ref<any> | string }> = (
  props: Props,
) => React.ReactNode;

type ExtractRefAttributesRef<T> = T extends React.RefAttributes<infer P> ? P : never;

/**
 * Get component ref
 * @example
 * ```ts
 * import { Input } from 'antd';
 * import type { GetRef } from 'antd';
 *
 * type InputRef = GetRef<typeof Input>;
 * ```
 * @since 5.13.0
 */
export type GetRef<T extends ReactRefComponent<any> | React.Component<any>> =
  T extends React.Component<any>
    ? T
    : T extends React.ComponentType<infer P>
      ? ExtractRefAttributesRef<P>
      : never;

export type GetContextProps<T> = T extends React.Context<infer P> ? P : never;

export type GetContextProp<
  T extends React.Context<any>,
  PropName extends keyof GetContextProps<T>,
> = NonNullable<GetContextProps<T>[PropName]>;
