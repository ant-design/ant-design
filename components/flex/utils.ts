import classNames from 'classnames';

import type { FlexProps } from './interface';

export const flexWrapValues: React.CSSProperties['flexWrap'][] = ['wrap', 'nowrap', 'wrap-reverse'];

export const justifyContentValues: React.CSSProperties['justifyContent'][] = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
];

export const alignItemsValues: React.CSSProperties['alignItems'][] = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
];

const genClsWrap = (prefixCls: string, props: FlexProps) => {
  const wrap = props.wrap === true ? 'wrap' : props.wrap;
  return {
    [`${prefixCls}-wrap-${wrap}`]: wrap && flexWrapValues.includes(wrap),
  };
};

const genClsAlign = (prefixCls: string, props: FlexProps) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach((cssKey) => {
    alignCls[`${prefixCls}-align-${cssKey}`] = props.align === cssKey;
  });
  alignCls[`${prefixCls}-align-stretch`] = !props.align && !!props.vertical;
  return alignCls;
};

const genClsJustify = (prefixCls: string, props: FlexProps) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach((cssKey) => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = props.justify === cssKey;
  });
  return justifyCls;
};

function createFlexClassNames(prefixCls: string, props: FlexProps) {
  return classNames({
    ...genClsWrap(prefixCls, props),
    ...genClsAlign(prefixCls, props),
    ...genClsJustify(prefixCls, props),
  });
}

export default createFlexClassNames;
