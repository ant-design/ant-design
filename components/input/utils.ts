import type { ReactNode } from 'react';

import type { InputProps } from './Input';

export function hasPrefixSuffix(props: {
  prefix?: ReactNode;
  suffix?: ReactNode;
  allowClear?: InputProps['allowClear'];
  showCount?: InputProps['showCount'];
}) {
  return !!(props.prefix || props.suffix || props.allowClear || props.showCount);
}
