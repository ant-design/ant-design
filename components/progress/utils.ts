import { presetPrimaryColors } from '@ant-design/colors';

import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({
  success = {},
  strokeColor,
}: Partial<CircleProps>): (string | Record<PropertyKey, string>)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
    strokeWidth?: number;
  },
): [number, number] => {
  let width = -1;
  let height = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    const strokeWidth = extra!.strokeWidth!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = strokeWidth ?? 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = (Array.isArray(size) ? size : [size.width, size.height]) as [
        number,
        number,
      ];
    }

    width *= steps;
  } else if (type === 'line') {
    const strokeWidth = extra?.strokeWidth;
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = strokeWidth || (size === 'small' ? 6 : 8);
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = (Array.isArray(size) ? size : [size.width, size.height]) as [
        number,
        number,
      ];
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else if (Array.isArray(size)) {
      width = (size[0] ?? size[1] ?? 120) as number;
      height = (size[0] ?? size[1] ?? 120) as number;
    }
  }
  return [width, height];
};
