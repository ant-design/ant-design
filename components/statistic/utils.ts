import type * as React from 'react';

export type valueType = number | string;
// countdownValueType is deprecated but still support
export type countdownValueType = number | string;

export type Formatter =
  | false
  | 'number'
  | 'countdown'
  | ((value: valueType, config?: FormatConfig) => React.ReactNode);

export interface FormatConfig {
  formatter?: Formatter;
  decimalSeparator?: string;
  groupSeparator?: string;
  precision?: number;
}

export interface CountdownFormatConfig extends FormatConfig {
  format?: string;
}

// Countdown
const timeUnits: [string, number][] = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function formatTimeStr(duration: number, format: string) {
  let leftDuration: number = duration;

  const escapeRegex = /\[[^\]]*]/g;
  const keepList: string[] = (format.match(escapeRegex) || []).map((str) => str.slice(1, -1));
  const templateText = format.replace(escapeRegex, '[]');

  const replacedText = timeUnits.reduce((current, [name, unit]) => {
    if (current.includes(name)) {
      const value = Math.floor(leftDuration / unit);
      leftDuration -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), (match: string) => {
        const len = match.length;
        return value.toString().padStart(len, '0');
      });
    }
    return current;
  }, templateText);

  let index = 0;
  return replacedText.replace(escapeRegex, () => {
    const match = keepList[index];
    index += 1;
    return match;
  });
}

export function formatCounter(value: valueType, config: CountdownFormatConfig, down: boolean) {
  const { format = '' } = config;
  const target = new Date(value).getTime();
  const current = Date.now();
  const diff = down ? Math.max(target - current, 0) : Math.max(current - target, 0);

  return formatTimeStr(diff, format);
}
