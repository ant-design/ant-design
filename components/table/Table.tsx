import * as React from 'react';
import { EXPAND_COLUMN, Summary } from 'rc-table';
import type { Reference } from 'rc-table';

import type { AnyObject } from '../_util/type';
import Column from './Column';
import ColumnGroup from './ColumnGroup';
import {
  SELECTION_ALL,
  SELECTION_COLUMN,
  SELECTION_INVERT,
  SELECTION_NONE,
} from './hooks/useSelection';
import type { RefTable } from './interface';
import type { TableProps } from './InternalTable';
import InternalTable from './InternalTable';

const Table = <RecordType extends AnyObject = AnyObject>(
  props: TableProps<RecordType>,
  ref: React.Ref<Reference>,
) => {
  const renderTimesRef = React.useRef<number>(0);
  renderTimesRef.current += 1;
  return <InternalTable<RecordType> {...props} ref={ref} _renderTimes={renderTimesRef.current} />;
};

const ForwardTable = React.forwardRef(Table) as unknown as RefTable & {
  displayName?: string;
  SELECTION_COLUMN: typeof SELECTION_COLUMN;
  EXPAND_COLUMN: typeof EXPAND_COLUMN;
  SELECTION_ALL: typeof SELECTION_ALL;
  SELECTION_INVERT: typeof SELECTION_INVERT;
  SELECTION_NONE: typeof SELECTION_NONE;
  Column: typeof Column;
  ColumnGroup: typeof ColumnGroup;
  Summary: typeof Summary;
};

ForwardTable.SELECTION_COLUMN = SELECTION_COLUMN;
ForwardTable.EXPAND_COLUMN = EXPAND_COLUMN;
ForwardTable.SELECTION_ALL = SELECTION_ALL;
ForwardTable.SELECTION_INVERT = SELECTION_INVERT;
ForwardTable.SELECTION_NONE = SELECTION_NONE;
ForwardTable.Column = Column;
ForwardTable.ColumnGroup = ColumnGroup;
ForwardTable.Summary = Summary;

if (process.env.NODE_ENV !== 'production') {
  ForwardTable.displayName = 'Table';
}

export default ForwardTable;
