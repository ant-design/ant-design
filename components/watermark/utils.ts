/** converting camel-cased strings to be lowercase and link it with Separator */
export function toLowercaseSeparator(key: string) {
  return key.replace(/([A-Z])/g, '-$1').toLowerCase();
}

export function getStyleStr(style: React.CSSProperties): string {
  return Object.keys(style)
    .map((key) => `${toLowercaseSeparator(key)}: ${style[key as keyof React.CSSProperties]};`)
    .join(' ');
}

/** Returns the ratio of the device's physical pixel resolution to the css pixel resolution */
export function getPixelRatio() {
  return window.devicePixelRatio || 1;
}

/** Whether to re-render the watermark */
export const reRendering = (mutation: MutationRecord, isWatermarkEle: (ele: Node) => boolean) => {
  let flag = false;
  // Whether to delete the watermark node
  if (mutation.removedNodes.length) {
    flag = Array.from<Node>(mutation.removedNodes).some((node) => isWatermarkEle(node));
  }
  // Whether the watermark dom property value has been modified
  if (mutation.type === 'attributes' && isWatermarkEle(mutation.target)) {
    flag = true;
  }
  return flag;
};
