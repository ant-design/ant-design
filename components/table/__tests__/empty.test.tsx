import React from 'react';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';

import type { ColumnsType } from '..';
import Table from '..';
import { render, triggerResize, waitFakeTimer } from '../../../tests/utils';

const columns: ColumnsType<any> = [
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
];

const columnsFixed: ColumnsType<any> = [
  {
    title: 'Full Name',
    width: 100,
    dataIndex: 'name',
    key: 'name',
    fixed: 'left',
  },
  {
    title: 'Age',
    width: 100,
    dataIndex: 'age',
    key: 'age',
    fixed: 'left',
  },
  { title: 'Column 1', dataIndex: 'address', key: '1' },
  { title: 'Column 2', dataIndex: 'address', key: '2' },
  { title: 'Column 3', dataIndex: 'address', key: '3' },
  { title: 'Column 4', dataIndex: 'address', key: '4' },
  { title: 'Column 5', dataIndex: 'address', key: '5' },
  { title: 'Column 6', dataIndex: 'address', key: '6' },
  { title: 'Column 7', dataIndex: 'address', key: '7' },
  { title: 'Column 8', dataIndex: 'address', key: '8' },
  {
    title: 'Action',
    key: 'address',
    fixed: 'right',
    width: 100,
  },
];

describe('Table', () => {
  it('renders empty table', () => {
    const { asFragment } = render(<Table dataSource={[]} columns={columns} pagination={false} />);
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  describe('renders empty table with fixed columns', () => {
    let domSpy: ReturnType<typeof spyElementPrototypes>;

    beforeAll(() => {
      domSpy = spyElementPrototypes(HTMLDivElement, {
        offsetWidth: {
          get: () => 1000,
        },
      });
    });
    afterAll(() => {
      domSpy.mockRestore();
    });

    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('should work', async () => {
      const { container, asFragment } = render(
        <Table dataSource={[]} columns={columnsFixed} pagination={false} scroll={{ x: 1 }} />,
      );

      triggerResize(container.querySelector('.ant-table')!);

      await waitFakeTimer();
      expect(container.querySelector('.ant-empty')).toBeTruthy();

      expect(asFragment().firstChild).toMatchSnapshot();
    });
  });

  it('renders empty table when emptyText is null', () => {
    const { container, asFragment } = render(
      <Table dataSource={[]} columns={columns} pagination={false} locale={{ emptyText: null }} />,
    );

    expect(container.querySelector('.ant-table-placeholder>.ant-table-cell')?.hasChildNodes()).toBe(
      false,
    );

    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('renders empty table with custom emptyText', () => {
    const { asFragment } = render(
      <Table
        dataSource={[]}
        columns={columns}
        pagination={false}
        locale={{ emptyText: 'custom empty text' }}
      />,
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });

  it('renders empty table without emptyText when loading', () => {
    const { asFragment } = render(<Table dataSource={[]} columns={columns} loading />);
    expect(asFragment().firstChild).toMatchSnapshot();
  });
});
