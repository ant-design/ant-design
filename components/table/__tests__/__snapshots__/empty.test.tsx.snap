// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders empty table 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table ant-table-empty"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
          >
            <table
              style="table-layout: auto;"
            >
              <colgroup />
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 1
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 2
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 3
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 4
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 5
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 6
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 7
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 8
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-placeholder"
                >
                  <td
                    class="ant-table-cell"
                    colspan="8"
                  >
                    <div
                      class="ant-empty ant-empty-normal"
                    >
                      <div
                        class="ant-empty-image"
                      >
                        <svg
                          height="41"
                          viewBox="0 0 64 41"
                          width="64"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <title>
                            No data
                          </title>
                          <g
                            fill="none"
                            fill-rule="evenodd"
                            transform="translate(0 1)"
                          >
                            <ellipse
                              cx="32"
                              cy="33"
                              fill="#f5f5f5"
                              rx="32"
                              ry="7"
                            />
                            <g
                              fill-rule="nonzero"
                              stroke="#d9d9d9"
                            >
                              <path
                                d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                              />
                              <path
                                d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                fill="#fafafa"
                              />
                            </g>
                          </g>
                        </svg>
                      </div>
                      <div
                        class="ant-empty-description"
                      >
                        No data
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table when emptyText is null 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table ant-table-empty"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
          >
            <table
              style="table-layout: auto;"
            >
              <colgroup />
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 1
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 2
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 3
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 4
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 5
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 6
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 7
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 8
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-placeholder"
                >
                  <td
                    class="ant-table-cell"
                    colspan="8"
                  />
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table with custom emptyText 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table ant-table-empty"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
          >
            <table
              style="table-layout: auto;"
            >
              <colgroup />
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 1
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 2
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 3
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 4
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 5
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 6
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 7
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 8
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-placeholder"
                >
                  <td
                    class="ant-table-cell"
                    colspan="8"
                  >
                    custom empty text
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table with fixed columns should work 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div
      class="ant-spin-container"
    >
      <div
        class="ant-table ant-table-empty ant-table-fixed-column ant-table-scroll-horizontal ant-table-has-fix-left ant-table-has-fix-right"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
            style="overflow-x: auto; overflow-y: hidden;"
          >
            <table
              style="width: 1px; min-width: 100%; table-layout: fixed;"
            >
              <colgroup>
                <col
                  style="width: 100px;"
                />
                <col
                  style="width: 100px;"
                />
                <col />
                <col />
                <col />
                <col />
                <col />
                <col />
                <col />
                <col />
                <col
                  style="width: 100px;"
                />
              </colgroup>
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-cell ant-table-cell-fix-left"
                    scope="col"
                    style="position: sticky; left: 0px;"
                  >
                    Full Name
                  </th>
                  <th
                    class="ant-table-cell ant-table-cell-fix-left ant-table-cell-fix-left-last"
                    scope="col"
                    style="position: sticky; left: 0px;"
                  >
                    Age
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 1
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 2
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 3
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 4
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 5
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 6
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 7
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 8
                  </th>
                  <th
                    class="ant-table-cell ant-table-cell-fix-right ant-table-cell-fix-right-first"
                    scope="col"
                    style="position: sticky; right: 0px;"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  aria-hidden="true"
                  class="ant-table-measure-row"
                  style="height: 0px; font-size: 0px;"
                >
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                  <td
                    style="padding: 0px; border: 0px; height: 0px;"
                  >
                    <div
                      style="height: 0px; overflow: hidden;"
                    >
                    </div>
                  </td>
                </tr>
                <tr
                  class="ant-table-placeholder"
                >
                  <td
                    class="ant-table-cell"
                    colspan="11"
                  >
                    <div
                      class="ant-table-expanded-row-fixed"
                      style="width: 1000px; position: sticky; left: 0px; overflow: hidden;"
                    >
                      <div
                        class="ant-empty ant-empty-normal"
                      >
                        <div
                          class="ant-empty-image"
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <title>
                              No data
                            </title>
                            <g
                              fill="none"
                              fill-rule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fill-rule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          class="ant-empty-description"
                        >
                          No data
                        </div>
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table renders empty table without emptyText when loading 1`] = `
<div
  class="ant-table-wrapper"
>
  <div
    class="ant-spin-nested-loading"
  >
    <div>
      <div
        aria-busy="true"
        aria-live="polite"
        class="ant-spin ant-spin-spinning"
      >
        <span
          class="ant-spin-dot-holder"
        >
          <span
            class="ant-spin-dot ant-spin-dot-spin"
          >
            <i
              class="ant-spin-dot-item"
            />
            <i
              class="ant-spin-dot-item"
            />
            <i
              class="ant-spin-dot-item"
            />
            <i
              class="ant-spin-dot-item"
            />
          </span>
        </span>
      </div>
    </div>
    <div
      class="ant-spin-container ant-spin-blur"
    >
      <div
        class="ant-table ant-table-empty"
      >
        <div
          class="ant-table-container"
        >
          <div
            class="ant-table-content"
          >
            <table
              style="table-layout: auto;"
            >
              <colgroup />
              <thead
                class="ant-table-thead"
              >
                <tr>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 1
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 2
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 3
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 4
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 5
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 6
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 7
                  </th>
                  <th
                    class="ant-table-cell"
                    scope="col"
                  >
                    Column 8
                  </th>
                </tr>
              </thead>
              <tbody
                class="ant-table-tbody"
              >
                <tr
                  class="ant-table-placeholder"
                >
                  <td
                    class="ant-table-cell"
                    colspan="8"
                  >
                    <div
                      class="ant-empty ant-empty-normal"
                    >
                      <div
                        class="ant-empty-image"
                      >
                        <svg
                          height="41"
                          viewBox="0 0 64 41"
                          width="64"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <title>
                            No data
                          </title>
                          <g
                            fill="none"
                            fill-rule="evenodd"
                            transform="translate(0 1)"
                          >
                            <ellipse
                              cx="32"
                              cy="33"
                              fill="#f5f5f5"
                              rx="32"
                              ry="7"
                            />
                            <g
                              fill-rule="nonzero"
                              stroke="#d9d9d9"
                            >
                              <path
                                d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                              />
                              <path
                                d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                fill="#fafafa"
                              />
                            </g>
                          </g>
                        </svg>
                      </div>
                      <div
                        class="ant-empty-description"
                      >
                        No data
                      </div>
                    </div>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
